### PR TITLE
Setting - Added multiline setting parsing support

### PR DESCRIPTION
**Pull request by johnb432**
**When merged this pull request will:**
- Lets `CBA_settings_fnc_parse` parse settings that are over multiple lines.
- The word `force` now needs to be followed by a space, tab, LF or CR in order for it to not be considered a setting.

E.g:
Previously:
```SQF
force forceforce = true; // Set the setting "force" to "be enforced 2x"
```
Now:
```SQF
force forceforce = true; // Sets the setting "forceforce" to "be enforced 1x"
```
- Addresses #1573 (which means [this](https://github.com/CBATeam/CBA_A3/wiki/CBA-Settings-System#userconfig-limitations) can be amended)

I have tested it, but not extensively.

### CODE REVIEW

**Comment by jonpas** [comment]
Wiki page amended.

File: addons/settings/fnc_parse.sqf [hunk 1]
```diff
@@ -13,65 +13,58 @@ Returns:
...
-    };
-
-    parseSimpleArray (["[", _string, "]"] joinString "") select 0
-};
-
 params [["_info", "", [""]], ["_validate", false, [false]], ["_source", "", [""]]];
 
-// remove whitespace at start and end of each line
-private _result = [];
+// Remove whitespace at start and end of each line, a line being define by the ";" at its end
```
Comment (commy2) [inline addons/settings/fnc_parse.sqf:21]: > a line being define by the ";" at its end

That's called "statement" not line.

File: addons/settings/fnc_parse.sqf [hunk 2]
```diff
@@ -13,65 +13,58 @@ Returns:
...
-// separate statements (setting = value)
-_result = [];
+// Separate statements (setting = value)
+private _result = [];
+private _indexEqualSign = -1;
+private _setting = "";
+private _priority = 0;
+private _value = "";
+private _countForce = count "force";
+private _whitespaces = [toString [ASCII_NEWLINE], toString [ASCII_CARRIAGE_RETURN], toString [ASCII_TAB], toString [ASCII_SPACE]];
```
Comment (commy2) [inline addons/settings/fnc_parse.sqf:38]:
```suggestion
private _whitespaces = toString [ASCII_NEWLINE, ASCII_CARRIAGE_RETURN, ASCII_TAB, ASCII_SPACE];
```
`in` (L48 & 54) can search strings.

File: addons/settings/fnc_parse.sqf [hunk 3]
```diff
@@ -13,65 +13,58 @@ Returns:
...
+    if (_setting select [0, _countForce] == "force" && {(_setting select [_countForce, 1]) in _whitespaces}) then {
+        _setting = (_setting select [_countForce]) trim [WHITESPACE, 1];
         _priority = _priority + 1;
     };
 
+    // If setting is valid, get it's value
     if (_setting != "") then {
-        private _value = ((_x select [_indexEqualSign + 1]) call CBA_fnc_trim) call _fnc_parseAny;
+        // Remove whitespaces; Parse bool, number, string
+        _value = parseSimpleArray (["[", (_x select [_indexEqualSign + 1]) splitString WHITESPACE joinString "", "]"] joinString "") select 0;
```
Comment (commy2) [inline addons/settings/fnc_parse.sqf:62]: idk, feels to me like this should be broken up.

File: addons/settings/fnc_parse.sqf (hunk 2 above)
Comment (commy2) [reply addons/settings/fnc_parse.sqf:38]: Even better, there is a WHITE_SPACE macro already. 

https://github.com/CBATeam/CBA_A3/blob/master/addons/strings/script_strings.hpp#L13

So just _whitespace = toString WHITE_SPACE ;

File: addons/settings/fnc_parse.sqf (hunk 3 above)
Comment (johnb432) [reply addons/settings/fnc_parse.sqf]: Reverted back to original formatting.

File: addons/settings/fnc_parse.sqf (hunk 2 above)
Comment (johnb432) [reply addons/settings/fnc_parse.sqf]: [WHITESPACE](https://github.com/CBATeam/CBA_A3/blob/c472830703a5c335bfa990c393a6b095d66d74bc/addons/settings/script_component.hpp#L149) already exists and is exactly what is wanted, so I used that instead.

File: addons/settings/fnc_parse.sqf (hunk 1 above)
Comment (johnb432) [reply addons/settings/fnc_parse.sqf]: Amended.

**Comment by johnb432** [comment]
Something that just came to mind is [forceUnicode](https://community.bistudio.com/wiki/forceUnicode), where is says [trim](https://community.bistudio.com/wiki/trim) supports the usage of `forceUnicode`.

I used this string: `"[Āā,Ăă,Ҙ,привет]"` to test. In my testing, `Ҙ` wasn't even recognised by Arma (I could not paste it into the console).

To test I added `forceUnicode = 0;` at the start of the function. However it didn't seem to make a difference: Both the old (the one in CBA v.3.15.8) and the new versions of `CBA_settings_fnc_parse` returned the same thing.

Ultimately, what I'm trying to convey is that I am slightly concerned about non-ASCII chars being a potential problem, but at the same time the small amount of testing I did suggests it's fine.

**Comment by rautamiekka** [comment]
> Something that just came to mind is [forceUnicode](https://community.bistudio.com/wiki/forceUnicode), where is says [trim](https://community.bistudio.com/wiki/trim) supports the usage of `forceUnicode`.
> 
> I used this string: `"[Āā,Ăă,Ҙ,привет]"` to test. In my testing, `Ҙ` wasn't even recognised by Arma (I could not paste it into the console).
> 
> To test I added `forceUnicode = 0;` at the start of the function. However it didn't seem to make a difference: Both the old (the one in CBA v.3.15.8) and the new versions of `CBA_settings_fnc_parse` returned the same thing.
> 
> Ultimately, what I'm trying to convey is that I am slightly concerned about non-ASCII chars being a potential problem, but at the same time the small amount of testing I did suggests it's fine.

Did you actually have a `forceUnicode = 0;` or a `forceUnicode 0;` ?

**Comment by johnb432** [comment]
> Did you actually have a `forceUnicode = 0;` or a `forceUnicode 0;` ?

No, that was a typo on my part, it was `forceUnicode 0;`.

**Comment by johnb432** [comment]
> > Did you actually have a `forceUnicode = 0;` or a `forceUnicode 0;` ?
> 
> No, that was a typo on my part, it was `forceUnicode 0;`.

Confirming that `trim` does not require `forceUnicode`.

**Review by jonpas — COMMENTED** [review]
Ready?

**Comment by johnb432** [comment]
> Ready?

Testing by others would be nice.

**Comment by Mike-MF** [comment]
Tested this with
```
ace_common_checkPBOsWhitelist = "[
    'ace_noactionmenu',
    'Blastcore_MainCore',
    'Blastcore_VEP',
    'WarFXPE',
    'ocap_main',
    'ocap_extension',
    'ocap_recorder'
]";
```

Came up in-game as
`force ace_common_checkPBOsWhitelist = "['ace_noactionmenu','Blastcore_MainCore','Blastcore_VEP','WarFXPE','ocap_main','ocap_extension','ocap_recorder']";`

it works.

**Comment by jonpas** [comment]
> it works.

In addition, it successfully parsed https://github.com/Theseus-Aegis/Mods/blob/96ce4f4a1c20029ad722a4cd992a314b461e6f7e/addons/cba_settings/cba_settings.sqf in its entirety as well (tested single-line and multi-line `ace_common_checkPBOsWhitelist`).

**Review by jonpas — APPROVED** [review]

**Pull request merged**